### PR TITLE
`InternalStreamConnection` must release buffers obtained via `ByteBufferBsonOutput.getByteBuffers`

### PR DESCRIPTION
**Pull request by stIncMale**
JAVA-5084

### CODE REVIEW

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java [hunk 1]
```diff
@@ -411,31 +412,28 @@ private void sendCommandMessage(final CommandMessage message,
...
+        List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
+        try {
+            if (localSendCompressor == null || SECURITY_SENSITIVE_COMMANDS.contains(message.getCommandDocument(bsonOutput).getFirstKey())) {
+                sendMessage(byteBuffers, message.getId());
+            } else {
                 CompressedMessage compressedMessage = new CompressedMessage(message.getOpCode(), byteBuffers, localSendCompressor,
                         getMessageSettings(description));
-                compressedBsonOutput = new ByteBufferBsonOutput(this);
-                compressedMessage.encode(compressedBsonOutput, sessionContext);
-            } finally {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:428]: Done.

File: driver-core/src/main/com/mongodb/connection/Stream.java [hunk 2]
```diff
@@ -46,7 +46,8 @@ public interface Stream extends BufferProvider{
     /**
      * Write each buffer in the list to the stream in order, blocking until all are completely written.
      *
-     * @param buffers the buffers to write
+     * @param buffers the buffers to write. The operation must not {@linkplain ByteBuf#release() release} any buffer from {@code buffers},
+     * unless it also {@linkplain ByteBuf#retain() retained} it, and releasing is meant to compensate for that.
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/Stream.java:50]: I think this is ok even in a minor release.  Just mark the JAVA ticket as Docs Changed Needed so we get something in the release notes.

**Review by jyemin — APPROVED** [review]
LGTM

File: driver-core/src/main/com/mongodb/connection/Stream.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/connection/Stream.java]: I documented the required doc change in the "Documentation Changes Summary" in https://jira.mongodb.org/browse/DOCSP-31709.

**Pull request merged**